### PR TITLE
HBO España + Nordic

### PR DESCRIPTION
**Pull request by Sakerdot**
https://github.com/arvvoid/plugin.video.hbogoeu/issues/6

Starting point for the development of HBO España.

ATM only https://es.hboespana.com/ login endpoint is supported, there isn't any info on episodes/seasons/tv shows but at least they can be played.

Code right now is messy and probably bug ridden, but if someone wanted to use HBO España on Kodi (specially after 15th of April...) here you go.

### CODE REVIEW

**Comment by Sakerdot** [comment]
It probably shouldn't be merged right now, I just wanted to have something to play HBO before the last GoT season is released.

**Comment by arvvoid** [comment]
Hello, good job :) I like the direction where this is going. Can't test the Spain handler because I don't have a Spain account but tested the add-on with your changes and it does not break any existing functionality, so it can be safely merged at any time.

For now, I will wait with merging until this is more mature and tested.

I made a zip for easy installation of your version so far, for easy testing: 
[plugin.video.hbogoeu-spain.zip](https://github.com/arvvoid/plugin.video.hbogoeu/files/3053190/plugin.video.hbogoeu-spain.zip)

File: hbogolib/handlersp.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+            'Accept-Language': self.ACCEPT_LANGUAGE,
+            'Accept-Encoding': 'gzip, deflate, br',
+            'Referer': self.API_HOST_GATEWAY_REFERER,
+            'Content-Type': 'application/xml',
+            'Authorization': 'Basic ' + base64.b64encode(username + ':' + base64.b64encode(password)),
+            'Origin': self.API_HOST_GATEWAY,
+            'Connection': 'keep-alive',
+        }
+
+        self.API_DEVICE_ID = self.generate_device_id()
```
Comment (arvvoid) [inline hbogolib/handlersp.py]: This way the device id will be generated on every add-on call my suggestion is to save device id to settings:

```
self.API_DEVICE_ID = self.addon.getSetting('individualization')

if (self.API_DEVICE_ID == ""):
    self.log("NO REGISTRED DEVICE - generating")
    self.API_DEVICE_ID = self.generate_device_id()
    self.addon.setSetting('individualization', str(self.API_DEVICE_ID))

```

File: hbogolib/handlersp.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        data = '<device><type>web</type><deviceId>' + self.API_DEVICE_ID + '</deviceId></device>'
+
+        response = self.send_login_hbogo(self.API_URL_AUTH_WEBBASIC, headers, data, 'xml')
+
+
+        if response.find('status').text == 'Success':
+            self.API_DEVICE_TOKEN = response.find('token').text
+            self.API_IDENTITY_GUID = response.find('identityGuid').text
+            self.API_ACCOUNT_GUID = response.find('accountGuid').text
+            self.init_api()
```
Comment (arvvoid) [inline hbogolib/handlersp.py:151]: Save sucessful login for reuse on next add-on calls:

```
 login_hash = hashlib.sha224(self.API_DEVICE_ID + username + password).hexdigest()
 self.log("LOGIN HASH: " + login_hash)
 saved_session = {

                "hash": login_hash,
                "API_DEVICE_TOKEN": self.API_DEVICE_TOKEN,
                "API_IDENTITY_GUID": self.API_IDENTITY_GUID,
                "API_ACCOUNT_GUID": self.API_ACCOUNT_GUID,
                "time": time.time()

}
self.save_obj(saved_session, self.addon_id + "_es_session")
```

File: hbogolib/handlersp.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+            'Accept-Encoding': 'gzip, deflate, br',
+            'Referer': self.API_HOST_GATEWAY_REFERER,
+            'Content-Type': 'application/xml',
+            'Authorization': 'Basic ' + base64.b64encode(username + ':' + base64.b64encode(password)),
+            'Origin': self.API_HOST_GATEWAY,
+            'Connection': 'keep-alive',
+        }
+
+        self.API_DEVICE_ID = self.generate_device_id()
+
```
Comment (arvvoid) [inline hbogolib/handlersp.py:142]: Before login check if there is a saved valid successful login:

```
login_hash = hashlib.sha224(self.API_DEVICE_ID + username + password).hexdigest()
self.log("LOGIN HASH: " + login_hash)

loaded_session = self.load_obj(self.addon_id + "_es_session")

if loaded_session != None:
    self.log("SAVED SESSION LOADED")
    if loaded_session["hash"] == login_hash:
      self.log("HASH IS VALID")
      if time.time() < (loaded_session["time"] + (self.SESSION_VALIDITY * 60 * 60)):
          self.log("NOT EXPIRED RESTORING...")
          self.API_DEVICE_TOKEN = loaded_session["API_DEVICE_TOKEN"]
          self.API_IDENTITY_GUID = loaded_session["API_IDENTITY_GUID"]
          self.API_ACCOUNT_GUID = loaded_session["API_ACCOUNT_GUID"]
          self.init_api()
          loaded_session['time'] = time.time()
          self.save_obj(loaded_session, self.addon_id + "_es_session")
          return True
```

**Comment by arvvoid** [comment]
@Sakerdot got confirmation that this work for HBO Nordic as well #7 with just modifying the URLs.

for spain is api-hbo**e**, for nordic api-hbo**n**   and self.API_HOST_GATEWAY = 'https://en.hbonordic.com/'

File: hbogolib/constants.py [hunk 4]
```diff
@@ -36,6 +36,7 @@ class HbogoConstants(object):
         ['Serbia', 'rs', 'sr', 'SRB', 'SRP', '', HANDLER_EU],
         ['Slovakia', 'sk', 'sk', 'SVK', 'SLO', '', HANDLER_EU],
         ['Slovenija', 'si', 'si', 'SVN', 'SLV', '', HANDLER_EU],
+        ['Spain', 'es', 'es', '', '', '', HANDLER_SPAIN],
```
Comment (arvvoid) [inline hbogolib/constants.py]: To add support for HBO Nordic as well here we need to modify and add:

```
        ['Spain', 'es', 'es', 'ESP', 'ESP', 'https://es.hboespana.com', HANDLER_SPAIN],
        ['Norway', 'no', 'no', 'NOR', 'NOR', 'https://no.hbonordic.com/', HANDLER_NORDIC],
        ['Denmark', 'dk', 'dk', 'DNK', 'DNK', 'https://dk.hbonordic.com/', HANDLER_NORDIC],
        ['Sweden', 'se', 'se', 'SWE', 'SWE', 'https://se.hbonordic.com/', HANDLER_NORDIC],
        ['Finland', 'fi', 'fi', 'FIN', 'FIN', 'https://fi.hbonordic.com/', HANDLER_NORDIC],
```

File: hbogolib/handlersp.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+import copy
+import xml.etree.ElementTree as ET
+
+import xbmcgui
+import xbmcplugin
+import inputstreamhelper
+
+class HbogoHandler_sp(HbogoHandler):
+
+    def __init__(self, handle, base_url):
```
Comment (arvvoid) [inline hbogolib/handlersp.py]: change to: (self, handle, base_url, country):

File: hbogolib/base.py [hunk 6]
```diff
@@ -54,6 +54,9 @@ def start(self):
         if HbogoConstants.countries[country_index][6] == HbogoConstants.HANDLER_EU:
             from hbogolib.handlereu import HbogoHandler_eu
             self.handler = HbogoHandler_eu(self.handle, self.base_url, HbogoConstants.countries[country_index])
+        elif HbogoConstants.countries[country_index][6] == HbogoConstants.HANDLER_SPAIN:
+            from hbogolib.handlersp import HbogoHandler_sp
+            self.handler = HbogoHandler_sp(self.handle, self.base_url)
```
Comment (arvvoid) [inline hbogolib/base.py]: change to: self.handler = HbogoHandler_sp(self.handle, self.base_url, HbogoConstants.countries[country_index])

File: hbogolib/handlersp.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        mpd_pre_url = media_item.find('.//media:content[@profile="HBO-DASH-WIDEVINE"]', namespaces=self.NAMESPACES).get('url') + '&responseType=xml'
+
+        mpd_url = self.get_from_hbogo(mpd_pre_url, 'xml').find('.//url').text
+
+        media_guid = media_item.find('.//guid').text
+
+        license_headers = 'X-Clearleap-AssetID=' + media_guid + '&X-Clearleap-DeviceId=' + self.API_DEVICE_ID + \
+            '&X-Clearleap-DeviceToken=' + self.API_DEVICE_TOKEN + '&Content-Type='
+    
+        license_url = 'https://api-hboe.hbo.clearleap.com/cloffice/drm/wv/' + media_guid + '|' + license_headers + '|R{SSM}|'
```
Comment (arvvoid) [inline hbogolib/handlersp.py]: change to:

`license_url = 'https://'+ self.API_HOST + '/cloffice/drm/wv/' + media_guid + '|' + license_headers + '|R{SSM}|'`

File: hbogolib/handlersp.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        self.API_DEVICE_TOKEN = ''
+        self.API_IDENTITY_GUID = ''
+        self.API_ACCOUNT_GUID = ''
+
+        self.NAMESPACES = {
+            'clearleap': 'http://www.clearleap.com/namespace/clearleap/1.0/',
+            'media': 'http://search.yahoo.com/mrss/',
+        }
+
+        self.API_HOST = 'api-hboe.hbo.clearleap.com'
```
Comment (arvvoid) [inline hbogolib/handlersp.py]: Change to support HBO Nordic with this handler:

```
        if country[1] == 'es':
            self.API_HOST = 'api-hboe.hbo.clearleap.com'
        else:
            self.API_HOST = 'api-hbon.hbo.clearleap.com'
        
        self.API_HOST_GATEWAY = country[5]
        self.API_HOST_GATEWAY_REFERER = self.API_HOST_GATEWAY + '/sign-in'

        self.API_URL_BROWSE = 'https://'+  self.API_HOST +'/cloffice/client/web/browse/'
        self.API_URL_AUTH_WEBBASIC = 'https://'+  self.API_HOST + '/cloffice/client/device/login'
```

File: hbogolib/handlersp.py [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+        vote_3 = (self.language(30723).encode('utf-8'), 'RunPlugin(' + self.base_url + "?url=VOTE&mode=8&vote=3&cid=" + content_id + ')')
+        vote_2 = (self.language(30724).encode('utf-8'), 'RunPlugin(' + self.base_url + "?url=VOTE&mode=8&vote=2&cid=" + content_id + ')')
+        vote_1 = (self.language(30725).encode('utf-8'), 'RunPlugin(' + self.base_url + "?url=VOTE&mode=8&vote=1&cid=" + content_id + ')')
+
+        if self.cur_loc == self.LB_MYPLAYLIST:
+            return [vote_5, vote_4, vote_3, vote_2, vote_1, remove_mylist]
+        else:
+            return [add_mylist, vote_5, vote_4, vote_3, vote_2, vote_1]
+
+    def generate_device_id(self):
```
Comment (arvvoid) [inline hbogolib/handlersp.py:92]: This can be changed to:

import uuid
return uuid.uuid4()

**Review by arvvoid — CHANGES_REQUESTED** [review]
Some minor changes that will make the session persistent, not make it generate a new device uuid on each request, add support for Nordic

**Review by arvvoid — APPROVED** [review]

**Comment by arvvoid** [comment]
New test version for Spain and Nordic please test and report back, please perform all actions and provide debug logs even if all seems to work fine.
[plugin.video.hbogoeu.test.spno.zip REMOVED]

#6 and #7 have been locked for conversation, from now on all conversation regarding HBO Spain and Nordic should be kept here until this is ready to merge.

**Comment by PolliSoft** [comment]
Thanks for the test version. Unfortunately it crashes on start, at login. See https://pastebin.com/izZUXkcZ for debug log.
File "C:\Users\Olof\AppData\Roaming\Kodi\addons\plugin.video.hbogoeu\hbogolib\handlersp.py", line 141, in login
                                                "time": time.time()
                                            NameError: global name 'time' is not defined

**Comment by arvvoid** [comment]
@PolliSoft tnx for reporting

Fixed version: 
[plugin.video.hbogoeu.test.spno.zip](https://github.com/arvvoid/plugin.video.hbogoeu/files/3094164/plugin.video.hbogoeu.test.spno.zip)

**Comment by PolliSoft** [comment]
Installed new version, and it works like a charm! No problem with sessions, all videos played as they should (tested around 20 of them).
Still no subtitles.

**Comment by macardi** [comment]
Hi, done some testing with the new version. It seems to work fine, no errors.

Logg: https://paste.ubuntu.com/p/3nnVnMdy65/

**Comment by jhdgghost** [comment]
Confirmed working hbo spain. But the subtitles are not working

**Comment by arvvoid** [comment]
Tnx for reporting.
Like i suspected subtitles are missing from the mpd manifest will have to be handled diferently.
Will get solved just need some time to work on it.
I will need a full debug log from Spain as well to work on the subtitles.

**Comment by Sakerdot** [comment]
Hey, great progress! I haven't had any time, hopefully I'll be able to help more soon.

**Comment by jhdgghost** [comment]
[01_KODI.log](https://github.com/arvvoid/plugin.video.hbogoeu/files/3096396/01_KODI.log)
Here one debug log from hbo spain. I hope i did correctly if not tell me and i try to do again. I check and there is audio options but not subs.

**Comment by arvvoid** [comment]
@jhdgghost It's fine, for the future you can follow the steps described here: https://kodi.wiki/view/Log_file/Easy

**Comment by sanangel** [comment]
@arvvoid ha, If there is any way to help so that the menus and titles of the films and series are in Spanish, let me know if I can help

**Comment by macardi** [comment]
Will movie, TV show and episode info and artwork, be supported in the future.

**Comment by arvvoid** [comment]
@sanangel  i think i figured out the languages, will implement it when i have a little spare time and then some testing will be needed for both Spain and Nordic. Also will need somebody to make language files.
Please take a look at ISSUE #3 for instructions and to submit  a translation.

@macardi yes soon i have the info necessary in the debug logs, but will take some trial and error probably.

**Comment by arvvoid** [comment]
It looks like to set the content language `?language=lang_code` have to be added to the api calls. Thats just by looking at the login page, so im not 100% sure but will try.

Language codes for Spain:

```
es_hboespana Spanish
en_hboespana English
```

Language codes for Nordic:

```
en_hbon English
da_hbon Dansk
no_hbon Norsk
fi_hbon Suomi
sv_hbon Svenska
```

**Comment by arvvoid** [comment]
Test release with language support and basic media info: 
[plugin.video.hbogoeu.test.spno.zip](https://github.com/arvvoid/plugin.video.hbogoeu/files/3100172/plugin.video.hbogoeu.test.spno.zip)

Please keep in mind I have no clue if this will work at all since I can't test it, try, and report back with full debug logs (follow these steps for the log https://kodi.wiki/view/Log_file/Easy). If it doesn't work post a log and revert to the previous version.

There is no language files yet so it will default to English no matter what your Kodi is set to.
To make read your local language from hbogo, go to the add-on configuration and turn on **Force default operator language**. This will have an effect only on the content retrieved from Hbo go but will not translate the add-on interface (messages, labels) for that I will need someone to do the translations as instructed in #3 

@PolliSoft @macardi @jhdgghost @sanangel

**Comment by jhdgghost** [comment]
I just tested and not working.

ERROR: GetDirectory - Error getting plugin://plugin.video.hbogoeu/
2019-04-20 15:48:45.127 T:1937490720   ERROR: CGUIMediaWindow::GetDirectory(plugin://plugin.video.hbogoeu/) failed

Revert to the previous one and working again.

**Comment by arvvoid** [comment]
How did you install the add-on did you use the install from zip function in kodi?
Please post full debug logs

**Comment by jhdgghost** [comment]
[01_KODI.log](https://github.com/arvvoid/plugin.video.hbogoeu/files/3100215/01_KODI.log)
Here the debug log. Yes i install from the zip function in kodi.

**Comment by macardi** [comment]
Im also get a login error. I did what I usually do, just downloded the zip-file and extracted it and pase it in the Kodi addon library after deleating the old addon folder.

Log: https://pastebin.com/mDp6GZ1R

I will give the translation for Swedish a go later this week.

**Comment by arvvoid** [comment]
Try this, it has basic media info, Spanish language file (contribution of @jhdgghost), support for local languages: 
[plugin.video.hbogoeu-esnotest.zip](https://github.com/arvvoid/plugin.video.hbogoeu/files/3100620/plugin.video.hbogoeu-esnotest.zip)

there is still a lot of room for improvement.

If your Kodi language is English or there is no language file in the add-on for your local language yet you have to enable Force default operator language in the add-on settings to make it retrieve content in your local language.

The recommended install method is using the install addon from zip function in Kodi. (in case it was previously installed in another way, removing the add-on and properly installing is recommended)

@PolliSoft @macardi @jhdgghost @sanangel @Sakerdot 

Work in progressSubtitles.

Things to add later on that i can't work on for lack of access:

- My List
- Context menus: Add to my list
- Context menus: Remove from my list
- Context menus: Voting
- Search

**Comment by sanangel** [comment]
@arvvoid ha If you need access, i can vive my id for prívate message

**Comment by arvvoid** [comment]
@sanangel tnx but no need, your account could get blocked for different geolocation access. I get the info I need to help from the debug logs. 

Does the last test version work for you, can you provide a full debug log?

**Comment by jhdgghost** [comment]
Just tested and is working fine. Nothing broken 👍

**Comment by sanangel** [comment]
> @sanangel tnx but no need, your account could get blocked for different geolocation access. I get the info I need to help from the debug logs.
> 
> Does the last test version work for you, can you provide a full debug log?

Yes, the latest version translated into Spanish and with information on the chapters and films works for me.
Although I miss some series of cartoons like Mundo Pocoyo.
Thanks for all your work.

**Comment by macardi** [comment]
Hi, at last I got some time away from the family to test the new version. It works very well. Info and art is there for the most part, missing info is year, genre, runtime as far as I can see. 

Log: https://paste.ubuntu.com/p/kSm8V3P5H7/

**Comment by jumakki** [comment]
Hi, thanks for great plugin!
I bumped into one issue which prevented login for me.
I got exception ending with:
```
File "/storage/.kodi/addons/plugin.video.hbogoeu/hbogolib/handler.py", line 252, in get_device_id_v1
                                                return hashlib.sha256(codecs.encode(str(mac), 'rot_13') + self.addon_id + '.credentials.v1.' + codecs.encode(str(space), 'rot_13')).digest()
File "/usr/lib/python2.7/encodings/rot_13.py", line 17, in encode
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 14: ordinal not in range(128)
```
My UI language is Finnish and 'space' contained a non-ascii character 'ä' in UTF-8: "227 MB Yhteensä"

Everything worked with Nordic after I hard-coded 'space' content to something else.

**Comment by arvvoid** [comment]
@jumakki Hello, tnx for reporting. A fix will be availible soon. If you have some spare time you can check #3 and submit a Finnish translation.

**Comment by arvvoid** [comment]
@jumakki Try please if this fix the login problem for you. It will ask you for credentials again, will not be able to decode your current credentials since your custom change. 
After you input credentials again it should work.

[plugin.video.hbogoeu-esnotest.zip](https://github.com/arvvoid/plugin.video.hbogoeu/files/3109918/plugin.video.hbogoeu-esnotest.zip)

**Comment by alfem** [comment]
Hi!  Another spanish user here.

I have tested the last version (https://github.com/arvvoid/plugin.video.hbogoeu/pull/13#issuecomment-485180020) but login shows an error notification. 

Curiosly I can access the films catalog, but no content get played.  

I can see this python error in my log:

https://pastebin.com/q4LjUjK8

Also tested the last .zip but it fails during installation.

**Comment by jumakki** [comment]
> @jumakki Try please if this fix the login problem for you.

Thanks! That fixed the problem.

**Comment by arvvoid** [comment]
@alfem please post a full debug log https://kodi.wiki/view/Log_file/Easy . Browsing partialy work without login.

**Comment by alfem** [comment]
Here you have the full log, after plugin deletion and fresh reinstall.

 https://paste.kodi.tv/isamatacuz.kodi

I hope it helps

**Comment by arvvoid** [comment]
@alfem do you by any chanse have a character outside the ascii range in you username or password? http://www.asciitable.com

**Comment by alfem** [comment]
Mmm, I am afraid not. Username is a gmail account with a dot in the middle (and the @ symbolf, obviously). And my probably unsafe password only contains simple ascii characters.

**Comment by arvvoid** [comment]
@alfem ok, will make a tentative fix in next version but not really sure whats the cause right now, since others dont report same problem.

Next version will include a workaround for missing subtitles as well. Subtitles seams are missing from manifest thats why dont get processed by inputstream. So probably hbo spain nordic handles them in some non standard way.  
So the addon will download the subtitle in local language separatly from hbo go convert it from ttml format to srt and set it in videoplayer, will take me some more days to finish that.

**Comment by Tr1aX** [comment]
I just updated to (through install from zip function in kodi) and tested the last version. On first try I got a login error notification, but worked well the second time I tried. 
Playback is perfect like before, and menus are in Spanish now with basic media info (artwork and episode synopsis) but maybe menu navigation is a bit slower now. Thanks for your work!

**Comment by arvvoid** [comment]
Test version with Swedish and Finnish language file and subtitles workaround (enable in add-on settings). This is not tested at all so it might not work right away please report any problems.

@alfem This contains a tentative fix for your problem as well. Remove the add-on and add-on data completely and install this version and try, please report the outcome.

[plugin.video.hbogoeu-esnotest.zip](https://github.com/arvvoid/plugin.video.hbogoeu/files/3124417/plugin.video.hbogoeu-esnotest.zip)

**Comment by jhdgghost** [comment]
I install from zip file option in kodi. The addon is working but not see any option in the settings for the subtitles.

**Comment by arvvoid** [comment]
Fix for the option, the option will be enabled by default: 
[plugin.video.hbogoeu-esnotest.zip](https://github.com/arvvoid/plugin.video.hbogoeu/files/3124748/plugin.video.hbogoeu-esnotest.zip)

**Comment by jhdgghost** [comment]
Just tested, the option now appears.But i test a couple of episodes and the subtitles not appear. Tell me if you need debug log

**Comment by jhdgghost** [comment]
An update. I tested game of thrones season 8 episode 2 and no subs but in s08E01 there is spanish subtitles. Tested also series The warrior none of the four episodes available have subtitles.

**Comment by jhdgghost** [comment]
[01_KODI.log](https://github.com/arvvoid/plugin.video.hbogoeu/files/3124783/01_KODI.log)
Here a debug log with working sub episodes and not working.

**Comment by arvvoid** [comment]
Try this version and please provide debug logs, also after playing something with subtitles in _.Kodi/userdata/addon_data/plugin.video.hbogoeu_  a folder named subs will get created containing the downloaded and converted subtitles please zip and send the content of the subs folder as well if possible.

[plugin.video.hbogoeu-esnotest.zip REMOVED BROKEN]

**Comment by PolliSoft** [comment]
Tested, and noticed the following things:
* No correct images for any media (Series/movies/episode thumbnails), only the default HBO image shown instead.
* Error when trying to play any media, debug log here: https://pastebin.com/AnXwT82j
2019-04-28 22:44:19.766 T:10716   ERROR: EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
                                             - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
                                            Error Type: <type 'exceptions.TypeError'>
                                            Error Contents: bad operand type for unary +: 'str'
                                            Traceback (most recent call last):
                                              File "C:\Users\Olof\AppData\Roaming\Kodi\addons\plugin.video.hbogoeu\main.py", line 24, in <module>
                                                addon_main.router(REQUEST_PARAMS)
                                              File "C:\Users\Olof\AppData\Roaming\Kodi\addons\plugin.video.hbogoeu\hbogolib\base.py", line 143, in router
                                                self.handler.play(url, content_id)
                                              File "C:\Users\Olof\AppData\Roaming\Kodi\addons\plugin.video.hbogoeu\hbogolib\handlersp.py", line 314, in play
                                                self.log("Manifest: " + + ET.tostring(mpd, encoding='utf8'))
                                            TypeError: bad operand type for unary +: 'str'
                                            -->End of Python script error report<--

**Comment by arvvoid** [comment]
@PolliSoft tnx, i done this in a hurry made some typos will fix it tomorrow.

**Comment by arvvoid** [comment]
Try this version: 
[plugin.video.hbogoeu-esnotest.zip REMOVED BROKEN]

**Comment by arvvoid** [comment]
Try this: 
[plugin.video.hbogoeu-esnotest.zip REMOVED BROKEN]

**Comment by PolliSoft** [comment]
Tested it, debug log here: https://pastebin.com/3iFGFrAW
No subtitles shown, but one Norwegian subtitle file (no_hbon.xml) downloaded to c:\Users\olof\AppData\Roaming\Kodi\userdata\addon_data\plugin.video.hbogoeu\subs\1f10ced-01092cdc642\. In the same folder, there is also a 0B no_hbon.srt, nothing else. 

I found this in the log:
2019-04-29 18:33:33.974 T:16096   DEBUG: [script.module.inputstreamhelper-0.3.4]: Found Widevine binary at C:\Users\Olof\AppData\Roaming\Kodi\cdm\widevinecdm.dll
2019-04-29 18:33:33.980 T:16096   DEBUG: [script.module.inputstreamhelper-0.3.4]: Widevine update check was made on 2019-04-18T12:08:51
2019-04-29 18:33:33.981 T:16096   DEBUG: [plugin.video.hbogoeu] Force subtitles enabled, downloading and converting subtitles in: C:\Users\Olof\AppData\Roaming\Kodi\userdata\addon_data\plugin.video.hbogoeu\subs/1f10ced-01092cdc642/
2019-04-29 18:33:33.981 T:16096   DEBUG: [plugin.video.hbogoeu] Processing subtitle language code: no_hbon URL: https://static.hbonordic.com/1f10ced-01092cdc642/35f2669b0aee41289756ff02bcc8a457.xml
2019-04-29 18:33:34.225 T:16096   DEBUG: [plugin.video.hbogoeu] Unexpected error in subtitles processing
2019-04-29 18:33:34.226 T:16096   DEBUG: [plugin.video.hbogoeu] Play url: <xbmcgui.ListItem object at 0x0000011268F82670>

Are the hard coded '/' in handlersp.py (line 341) the problem for Windows here?

Is there anything else I can upload to help?

**Comment by jumakki** [comment]
Hard coded '/' are not the problem.
I added some debug logging and it fails in Ttml2srt::extract_dialogue() when it bumps into non-ascii character.
I tested with GoT episode 3 and nordic languages fail quite soon,
I skipped others than English but it fails too when it gets this line:
`<p xml:id="14" xml:space="preserve" begin="00:09:03:07" end="00:09:04:18" tts:backgroundColor="black" tts:fontSize="36px" tts:origin="0px 477px"><span tts:fontStyle="italic">Valar morghulis.</span></p>`
It is first text line with that span italic.

**Comment by jumakki** [comment]
English with that span failed because there was "self." missing  from this line in ttml2srt.py around line 30 when calling recursively extract_dialogue():
`            
if node.hasChildNodes():
                dialogue = dialogue + self.extract_dialogue(node.childNodes)
`

**Comment by PolliSoft** [comment]
Thanks a lot for this, @jumakki. I did the change and all subs are now working!

**Comment by arvvoid** [comment]
@jumakki @PolliSoft tnx for your contributions and testing. Fixed it.

This version should do the trick let me know the outcome.

[plugin.video.hbogoeu-esnotest.zip](https://github.com/arvvoid/plugin.video.hbogoeu/files/3129176/plugin.video.hbogoeu-esnotest.zip)

**Comment by jhdgghost** [comment]
Tested. And the everything is working perfectly. Several episodes tested and english and spanish subs are available. Thanks for the fix

**Comment by jumakki** [comment]
@arvvoid Thanks, works perfectly!

**Comment by macardi** [comment]
Just tested the new version, works very well also at my end. Thanks a lot!

**Comment by arvvoid** [comment]
Now remains to see how to get my list, and the api call to add to my list and remove from my list, and the api call to vote.

After that, this is ready to be merged, and included in the standard releases.

**Comment by sanangel** [comment]
Thanks @arvvoid ha for all your work 
The last version work perfect.

**Comment by jhdgghost** [comment]
A little question the version of this branch is lower than of the main branch. As I have installed the repository an the update notification is always there and i have to enter in the addon section to confirm if there is another addon update and can not select update all if i do not want hbogoeu addon get updated to last main branch. 

Can I manually change the version so it is the same than the main branch so the update notification for this addon disappears? Disabling the repository not work

I know is it an stupid thing but I am curious to know if i can do it myself and also happens to me previously with Netflix addon for example.

**Comment by arvvoid** [comment]
@jhdgghost you can simply disable automatic updates for this add-on untill it get merged. If you want to manualy change the version to the same as the main branch nothing bad will happen.

**Comment by alfem** [comment]
> @alfem This contains a tentative fix for your problem as well. Remove the add-on and add-on data completely and install this version and try, please report the outcome.

It works! 

Well, sound is in English (but menus and titles are in Spanish). I do not know if this is the expected behaviour.

We are reaaally close! Thanks for your amazing work!

**Comment by arvvoid** [comment]
@alfem Glad it works! 
What audio tracks are availible depends on the content and what hbo provide and specify in the manifest. You can change beetween availible audio streams in the audio menu during playbeck, thats beyond the add-on control, get handled by inputstream.adaptive.
To set default audio and subtite language go to Kodi Settings -> appearance -> international you should be able to select your default audio and subtitle language (will get autoselected if availible in the stream).

**Comment by jhdgghost** [comment]
[en_hboespana.en-US.txt](https://github.com/arvvoid/plugin.video.hbogoeu/files/3147194/en_hboespana.en-US.txt)

A little issue i find in last episode of game of thrones 8x04. After 1 hour the subs dissapear. I upload the sub. It is clear that the problem is that the time is wrong because the minutes are not reset to zero after one hour. Happens the same in english and spanish subs. Not sure if it is addon error or is a problem in the subs downloaded. Also check game of thrones 8x03 and happens the same. Seems a problem with more than one hour episodes.

**Comment by arvvoid** [comment]
@jhdgghost its probably a bug in the ttml to srt converter will look into it. Please post the original subs in ttml format as well so i can debug the converter

**Comment by alfem** [comment]
> @alfem Glad it works!
> What audio tracks are availible depends on the content and what hbo provide and specify in the manifest. You can change beetween availible audio streams in the audio menu during playbeck, thats beyond the add-on control, get handled by inputstream.adaptive.

That made the trick! 
I had got kodi with language setting as 'default' and English is the default for almost everything. I did not notice this because my kodi interface is in Spanish.

I hope you could merge this branch soon. Thanks for your help!

**Comment by jumakki** [comment]
Counting minutes has a bug in ms_to_subrip() in ttlm2srt.py
Just int(ms / 60000) isn't enough. Would need to substract hours * 60 from it.
It worked when changed to this:
```
    def ms_to_subrip(self, ms):
        return '{:02d}:{:02d}:{:02d},{:03d}'.format(
            int(ms / (3600 * 1000)),  # hh
            int(ms / 60000 - (ms / (3600 * 1000) * 60)),  # mm
            int((ms % 60000) / 1000),  # ss
            int((ms % 60000) % 1000))  # ms
```

**Comment by jhdgghost** [comment]
The subs in xml. 
[en_hboespana.txt](https://github.com/arvvoid/plugin.video.hbogoeu/files/3149683/en_hboespana.txt)

**Comment by arvvoid** [comment]
Fixed subtitles version: 
[plugin.video.hbogoeu-esnotest.zip](https://github.com/arvvoid/plugin.video.hbogoeu/files/3151094/plugin.video.hbogoeu-esnotest.zip)

Tnx @jumakki and tnx @jhdgghost for reporting.

I'm planning of merging this soon will leave it for couple more days, i think this is ready to be used and tested by more people now.
Then the missing features SEARCH, MY LIST, ADD/ROMOVE FROM MY LIST AND VOTING can be added with new pull request/s.
After merging i will start working on the kodi library integration feature.
What you all think about that? @Sakerdot @jumakki

**Comment by sanangel** [comment]
It works perfectly, although some options are missing, it is more than suitable for daily use.

**Comment by Sakerdot** [comment]
I've tried it a little bit and it seems fine, I'll have time again by the end of June, hopefully I can help then.

**Comment by jumakki** [comment]
I agree. It's ready for merging

**Comment by arvvoid** [comment]
@all-contributors please add @Sakerdot  for code, ideas

**Comment by arvvoid** [comment]
@all-contributors please add @PolliSoft for bug, test

**Comment by arvvoid** [comment]
@all-contributors please add @PolliSoft for bug, test

**Comment by arvvoid** [comment]
@all-contributors please add @macardi for translation, test

**Comment by arvvoid** [comment]
@all-contributors please add @jhdgghost for translation, test

**Comment by arvvoid** [comment]
@all-contributors please add @jumakki  for translation, test, bug, ideas

**Comment by arvvoid** [comment]
@all-contributors please add @sanangel  for testing

**Comment by arvvoid** [comment]
@all-contributors please add @alfem for testing

**Pull request merged**